In [1]:
import math

### (a). Using the partition function and β = 0.5 compute the E-step and show the HiddenMatrix.

In [19]:
data_j = [[0.1, 0.2],[0.2, 0.1],[0.3, 0],[1, 1.2],[0.8, 1],[9, 0.1]]
cluster_c = [[0.1, 0.9], [0.5, 0], [0.9, 0.5]]
beta = 0.5

def _partition_function(beta, data, center):
    return math.e**(-beta * _find_distance(data, center))

def _find_distance(point1, point2):
    return math.sqrt((point1[0]-point2[0])**2 + (point1[1] - point2[1])**2)

def _sum(data, center, beta):
    sum = 0
    for i in center:
        sum += _partition_function(beta=beta, data=data, center=i)

    return sum

def _dot_product(a, b):
    dot_product=0
    for a,b in zip(a,b):
        dot_product += a*b

    return dot_product

def _E_step_a(data, center, beta):
    hidden_matrix = [[0]*6 for i in range(3)]

    for i in range(len(center)):
        for j in range(len(data)):
            sum = _sum(data[j], center, beta)
            hidden_matrix[i][j] = _partition_function(beta, data[j], center[i]) / sum


    return hidden_matrix

hidden_matrix = _E_step_a(data_j, cluster_c, beta)
hidden_matrix



[[0.3267511728523752,
  0.3050994280684201,
  0.28508404512158053,
  0.33700778468570974,
  0.33914505262395894,
  0.26634685780600265],
 [0.3707737892616577,
  0.3898011438631598,
  0.4090188548359777,
  0.28271736237522366,
  0.28656516573360846,
  0.33110868964673695],
 [0.3024750378859669,
  0.30509942806842016,
  0.30589710004244164,
  0.3802748529390666,
  0.3742897816424327,
  0.4025444525472604]]

### (b). Using the assignments you made in (a), compute the M-step.

In [20]:
def _M_step_a(hidden_matrix, data):
    x_i1 = []
    x_i2 = []
    uni_vec = [1] * 6
    center = [[0] * 2 for i in range(3)]

    for i in data:
        x_i1.append(i[0])
        x_i2.append(i[1])

    for i in range(len(center)):
        center[i][0] = _dot_product(hidden_matrix[i], x_i1) / _dot_product(hidden_matrix[i], uni_vec)
        center[i][1] = _dot_product(hidden_matrix[i], x_i2) / _dot_product(hidden_matrix[i], uni_vec)

    return center

_M_step_a(hidden_matrix=hidden_matrix, data=data_j)


[[1.7127067587075437, 0.46575952602088455],
 [1.8017961268475913, 0.37298421952578614],
 [2.1663697397259987, 0.464545536710244]]

### (c). Using the Newtonian inverse-square law of gravitation, compute the E-step.

In [24]:
def _partition_function_c(data, center):
    return 1/_find_distance(data, center)**2

def _sum_c(data, center):
    sum = 0
    for i in center:
        sum += _partition_function_c(data=data, center=i)

    return sum

def _E_step_c(data, center):
    hidden_matrix = [[0]*6 for i in range(3)]

    for i in range(len(center)):
        for j in range(len(data)):
            sum = _sum_c(data[j], center)
            hidden_matrix[i][j] = _partition_function_c(data[j], center[i]) / sum


    return hidden_matrix



hiddenMatrix = _E_step_c(data_j, cluster_c)
hiddenMatrix

[[0.24264583679574545,
  0.11764705882352938,
  0.04229502513433871,
  0.3000710227272727,
  0.29570116861435725,
  0.301284974459967],
 [0.5944823001495761,
  0.7647058823529412,
  0.8987692841046975,
  0.15980113636363633,
  0.13564273789649417,
  0.3329311543125985],
 [0.16287186305467838,
  0.11764705882352944,
  0.05893569076096377,
  0.5401278409090909,
  0.5686560934891486,
  0.36578387122743455]]

### (d). Using the assignments you made in (c), compute the M-step.

In [25]:
def _M_step_a(hidden_matrix, data):
    x_i1 = []
    x_i2 = []
    uni_vec = [1] * 6
    center = [[0] * 2 for i in range(3)]

    for i in data:
        x_i1.append(i[0])
        x_i2.append(i[1])

    for i in range(len(center)):
        center[i][0] = _dot_product(hidden_matrix[i], x_i1) / _dot_product(hidden_matrix[i], uni_vec)
        center[i][1] = _dot_product(hidden_matrix[i], x_i2) / _dot_product(hidden_matrix[i], uni_vec)

    return center

_M_step_a(hidden_matrix=hiddenMatrix, data=data_j)

[[2.5458329085157407, 0.5741634960914257],
 [1.2980888076713388, 0.1926542649393067],
 [2.395011670363753, 0.7153864004819717]]

### (e) Any observations comparing the two different distance functions?

As I can observe from the updated centers for two different distance functions, the first center and the third center points stayed more or less the same for both method. The big difference is that for the second cluster center, it was (0.44, 0.39) with one method and now becomes (0.27, 0.09) with the other one.